src/Std/Time.lean

```diff
@@ -0,0 +1,173 @@
+
+## Date and time
+Combines date and time into a single representation, useful for precise timestamps and scheduling.
+
+- **`PlainDateTime`**: Represents both date and time in the format `YYYY-MM-DDTHH:mm:ss.sssssssss`.
+- **`Timestamp`**: Represents a point in time with nanosecond-level precision. It starts on the UNIX
+epoch and it should be used when you receive or need to send timestamps to other systems.
+
+## Zoned date and times.
+Combines date, time and time zones.
+
+- **`DateTime`**: Represents both date and time but with a time zone in the type constructor.
+- **`ZonedDateTime`**: An existential version of the `DateTime`.
+
+## Duration
+Represents spans of time and the difference between two points in time.
+
+- **`Duration`**: Represents the time span or difference between two `Timestamp`s values with nanosecond precision.
+
+# Formats
+
+Format strings are used to convert between `String` representations and date/time types, e.g, `YYYY-MM-DD'T'hh:mm:ss.sssZ`.
+The supported formats include:
+
+- **Year:**
+  - `YYYY`: Four-digit year (e.g., 2024)
+  - `YY`: Two-digit year (e.g., 24 for 2024)
+- **Month:**
+  - `MMMM`: Full month name (e.g., January, February)
+  - `MMM`: Abbreviated month name (e.g., Jan, Feb)
+  - `MM`: Two-digit month (e.g., 01 for January)
+  - `M`: One or two-digit month (e.g., 1 for January, 10 for October)
+- **Day:**
+  - `DD`: Two-digit day of the month (e.g., 01, 02)
+  - `D`: One or two-digit day of the month (e.g., 1, 2)
+  - `d`: One or two-digit day of the month with space padding (e.g., " 1", "12")
+- **Day of Week:**
+  - `EEEE`: Full name of the day of the week (e.g., Monday, Tuesday)
+  - `EEE`: Abbreviated day of the week (e.g., Mon, Tue)
+- **Hour:**
+  - `hh`: Two-digit hour in 24-hour format (e.g., 13, 14)
+  - `h`: One or two-digit hour in 24-hour format (e.g., 1, 2)
+  - `HH`: Two-digit hour in 12-hour format (e.g., 01, 02)
+  - `H`: One or two-digit hour in 12-hour format (e.g., 1, 2)
+- **AM/PM Indicator:**
+  - `AA`: Uppercase AM/PM (e.g., AM, PM)
+  - `aa`: Lowercase am/pm (e.g., am, pm)
+- **Minute:**
+  - `mm`: Two-digit minute (e.g., 01, 02)
+  - `m`: One or two-digit minute (e.g., 1, 2)
+- **Second:**
+  - `sssssssss` : Nine-digit nanosseconds (e.g., 000000009)
+  - `sss`: Three-digit milliseconds (e.g., 001, 202)
+  - `ss`: Two-digit second (e.g., 01, 02)
+  - `s`: One or two-digit second (e.g., 1, 2)
+- **Timezone:**
+  - `ZZZZZ`: Full timezone offset with hours and minutes (e.g., +03:00)
+  - `ZZZZ`: Timezone offset without the colon (e.g., +0300)
+  - `ZZZ`: Like `ZZZZ`, but with "UTC" for UTC
+  - `Z`: Like `ZZZZZ`, but with "Z" for UTC
+  - `z`: Timezone name (e.g., Brasilia Standard Time)
+
+# Macros
+
+In order to help the user build dates easily, there are a lot of macros available for creating dates.
+The `.sssssssss` can be ommited in most cases.
+
+- **`date% YYYY-MM-DD`**: Defines a date in the `YYYY-MM-DD` format.
```

Comment:
Is `date%` supposed to be ISO 8601 conform? I went through the bug tracker of https://github.com/time-rs/time and https://github.com/chronotope/chrono to look for some parsing bugs etc. that might show interesting edge cases and there are tickets that contain string `date%` cannot parse but their 8601 parser can parse, for example:
```lean
#eval date% 2022-07-22T12:52:50.349409+00:00 -- https://github.com/time-rs/time/issues/488
#eval date% 2023-12-11T11:31:60.000Z -- https://github.com/time-rs/time/issues/674
#eval date% 2017-06-22T13:32:16Z -- https://github.com/chronotope/chrono/issues/163
```
Are these in scope?

Reply:
I think that these are not supposed to be compliant, mainly because it seems we cannot use `.` in our notation for parsing reasons? It might be a good idea to change the notation from separating date and time with a colon to separating them with a `T` instead.

Reply:
I also looked in ISO 8601:2004 (I haven't got hold of ISO 8601:2019 yet, but I might later), and there for the decimal separator they allow `.` and `,` and comma is preferred over period. Colon is disallowed. So one could argue that the best format would be `date% YYYY-MM-DDTHH:mm:ss,sssssssss`

Reply:
I'm having problems trying to create a syntax with `noWs 'T' noWs`, the `quotation` gets confused with something like `$date:dateT$time:time`. The other problem is that the `sssssssss` should be the fractional part, not the nanoseconds part, so if someone writes `13:12:11,1` the 1 will be understood as `0000000001` nanoseconds.

Reply:
> the `quotation` gets confused with something like `$date:dateT$time:time`.

`$(date)T$(time)` seems to work.

Reply:
I tried using your suggestion with `syntax date noWs "T" noWs time : datetime`, but I'm still running into issues. specifically, when I changed [this line](https://github.com/leanprover/lean4/pull/4904/files#diff-331a2ce2fba831c8b23c02b797c734d6204f36a3deb416cf69239039e69303edR107) line to `(datetime| $(date)T$(time))`, I got an `unexpected token 'T'; expected ')'` error.